### DISCUSSION TITLE
Make it easy to opt out of next/babel preset components

### DISCUSSION THREAD

**Discussion by swyxio** [vercel/next.js#23922] [Ideas]
# Feature request

## Is your feature request related to a problem? Please describe.

Sometimes the provided babel plugins conflict with what I want to do. For example, I would like to [use `class` instead of `className` and Preact with Next.js](https://dev.to/swyx/how-to-use-class-instead-of-classname-with-preact-and-typescript-2bjh). 

However, `styled-jsx`'s babel plugin is overwriting my classes:

![image](https://user-images.githubusercontent.com/6764957/78468866-23343580-774e-11ea-8c3a-e05701579a5b.png)

since I use `Tailwind` instead of `styled-jsx`, I would like to turn the styled-jsx plugin off.

There is an old issue to [opt out of styled-jsx](https://github.com/zeit/next.js/issues/2384), but this no longer works since plugins are no longer self identifying:

[GitHub Source](https://github.com/zeit/next.js/blob/canary/packages/next/build/babel/preset.ts) 

You cant simply do `myPreset.plugins.filter` any more because there is no identifying name for the plugins. You can't rely on order because the order varies depending on env.

in order to do what i need i am having to [copy out the entire source of the babel plugin](https://prettier.io/playground/#N4Igxg9gdgLgprEAuEAdEBXAznABFmAJwEswZ0BuSKA3BANwF4AHQiMOLLAOge4DkA8gBEAogH1R-AGpVotYlgAKbACYYyxaIwaM9AclYR1m6Prk0YuRcLj04AGwjMAtghg6oTA6ruPnbrDm1ApYACqcHroG8ATmAPTxuABKnBAO9vgwAJ4OcKoAtABWWAAeuMwOGADmxDSoUABmGFCmUFm5+QBSZYLMMFo0ABTOA-IAlMDEjUMAhKODWJOEcDAYhFDAAL4UW9NzAIKEhACG2dyKR6fZI-2L3JU1dUvLq+vtC-K7nzQPVbU0Rg-HiPAE8FwnZhDUF1RgAPimMyuZwuWGRNxhUHGkxCMAA2lATm4ADTAgC6LH+dQoKzWGzxKwAjhhiCtuCssOl7ENCW5xqS7vIybtae9cEyWWyOVy4NCqVjduMaW8NrhgbtErgwicANYIXCNNguXAACxgMGYWCQiVqMBNGAARtxIC54g6Tg7HG6PV6XS5iDB4qoAGwABjAAFY4ABGADsAGZgxGACzxk7BuAnMDJh0pgCcEYj8ZjodjJzz7rzwdjoYj0YAxKppo0Cg6c3BM6oABwnVR56MFxPRttF2Nh6NgMAAJinRanyWjEa7DWarTG7SwGGYzAghBgWAAyjATgMwKIDwBZIZgE4OPKEV50qCzWbX2-3vSMKAYO8AH1-N53nAhCfvQEDEKouChgA-GBEFQUggH3twm7bru+5HiepDnheSpbC4xg-nAvClDue5YIwQyQsQArrhR2zjPCwD0CchC4OIwIACSsJwqyiF4Fi0KhZH7jhjDUc677AUMwnoYex6njhSq4tYh7AfYIESUh0naSBcIvrpABkhm6aiB7qcBynyFYigXsYwFQOJzDEJJQGEG+bnwgZUmEMZpm2fZGxWZYqkADInKUdm+BsjABdFUD-kMQwcYKNDcRyfFeECqVYHihgZTABQMPoZKMXo35-r+KV0elvEwPx3iMHBkEwc1CHVYstU4PVXjcMehDVKsSzGcCeU8d1RVeCVfWsYN+68FgBHqHkFF6EQGBwIJVjjZl9AAMLQI0xDVIwwCagAku0dp4LEVgMKy0CBFYAAGS1EVgz2qWqjTwO0UAdr4kEwBAuCevgqy4MDkOEBtxKgz6DgGsd6ycJDJonmqGBWA62TWPujiNLg2B1NUaN4M9+gnFjED6J9PwNBd7RGCY65Bn4TiuO4aOKGqqVfd1kMg89jS3jgn2cmjGMAO5wA6zBZjquA3u06NQKoeTWC4InxHApHobgUsBia0MdgUWDozqJMNG9K1IJT1P6MSutgFUvhIHlRAnDQjS7i4BQ5MwcAQC2WDZC4DrpCVxLcDHo35XVk30CVOyagA6ia+qezQzkOCTBq7mT4OEBpapsbdWBw-1c2F2A6wrLAuD8PZuAaVggwM0k0y4FAEB3aRudgAGDi41ggeD0d+RIA0+xDIo5nF8B-6KBEBDjMZcw7T1+2HcdM0DUN-6vgHQeE5vDUHU0u9V0Nn76BADpFHAZD6MZ+g974+h1BUBXnzv1R73NF4kwz5eAvkdf+199ynU1Nqfe+4a51y5kYDgXB9At2Am3aA31G72W4FqE0PMJSslRpCS0oM4Ad1BiyBwVgvaQV8JUCA2R8hqiuhnfARI8Ct0GNgpuvhuANHfnAO2tdjjuH0DsLYmpR5P2mLjE4386qCyJjgawlhOzYIAALuk9A4eIm9E4NBUrXAgEAXBRQcioOqsUsDhUioFKAI0cp5V1oHEgT1bwFCWg5Aom8SrKifMATk6wOBhGyIHO2LQiQOmRhAbAjtN5WjxCY4G5iHFWO6v+BkcBmTEKGPobRCN9EFUTvocY3BfAix-DAYkIDt6X2qGSYk2TckrHyYU3RxSE4rCzDAMpxIzpJDCAQrAuBeyqFGR0r0mJ-anG9r7AoPSyDFDKGbOJhAOBjLVpQqZeiZlZywD7QgfslmFRKKUM2BMxnUwhKed8uMv6+HsBzJ6DQnn+E5rAJANh2YBHcEvcIkRamnGqBCO24hxDnMdjHbgccDGnOTk0lpkpZQFJ0dMkpx8sBgBIP0fpwB3wQClvwTho8sycCQOtOAWwyRNMxEk5FeT9DcH0fKLA8Rzm+JBRCfFMCRnf0IigwuPs7xErznUR+ZAWHEC1vrI5uAbao2oMeOoecugHgABpTygJqXAmsRIpEzGQA0RpcD6ARRQSheqWJsQhec3AjBDW9OdD0+Aog8hPUtVARVdsEXEhlSJA4Vp9CpF6Qk7lJxwWQrKOG5wwEcgiNdXAd1cAnoSKRUQtpzLWVPBoPEBY-oABecACgmggBAHUBRfAEBhmQdYJNeVJGgMPXmAwi14DLRWg0LQ2ijIDehFhhozFOrIPEKxvSGi5wdJSmG1KmmZtRbsnNAIzbZFgBFKta6iSkAKP2vc-SF35JZfS+IpyChOF7B6PIvj5T9MZVmpdMyjA7iwJ4l2JwuBcrjXuYgnB+lx3fZ+598bf1YBKv+bYGackovaeivZ8ov0vs8ffSVhUOSFVHj01Q+LsBwAAELUJgJdK0VKaXElmHPCyvlDL3sXXB5ddRZle0OQsmGsAZVwHxZAFYJQkAiwcDgYkGcHBuJI7O4kKxBr-VOMDQgM7Ya4ZwoRW2slyI4WMnUsBu9FVYFmAYP00ASiOw9JyKo8BkgtDbcI5BnAeDcPkA8KAzBoIQqbBsThSAWiVLqPkaOsdnH6AOUck5lmOOIuacvSITi6J5QIJ0QoUKyTRcWLFnIeQEsxrJHlOD-tIglWgoe-QcX0srNKN6XRuW4jjCQIV4r+RSvlccP0urqgeilD6HRW4MWitpfq4l8Y87oNMuPWy+IRJmAFBPEQYgDosZ-v5IoFQgr1zGVo-knL+y5kseOYso1aHU0QHsIh-24T5sDJWARew51ZV7nk3OiTQ2H30afWwJDDgCg-DfejVVUBqgHse3RopL3nAQFfe9iqudzYFEgLeTgg9fsfbcSeXc-TKML2o4Vx9CHQ7rouTEgEfT+RrbRUDhDz7QeeO-G4EgYBLnyxkyj8plSqY0OJPxnAtLuBHRodJPD5a8he35Id4CJBq3u2ALda0qBUB9TKNBTi8Ralsvdpj57ZPXsU-BxganO6cD0+R4QMpFS4BVJoWSGlkiKAgGJCAYEyBQCsTYFLJQrEEBYGQCAW8Uszju5tw6U4YA9QwAPPLeH1RkBUpt2aFwDh04Bk4KHuA5kaABmIPQAM2QPdgC4NbkAzx40qBOKCk4yB2dwBt+cvDAeg8Hk4aFHzpfRbl5AOcg8JM8gAEUMC9zgI3gTzf6c4EIB7uDufWB1BgKnCCdpkBdlDDbowOBU6nGYB7naC9c-Mh70ty0HuP0FH+vkfIueF2F+L33wTIAcD+gj+Jq-7e4Bd57xf5vx4HRT9UDPpAU4bee2ILnX7A6FwCED3TgCMXPXDbUB0d3PjJvG3egDaS6XwWAA8HFYgfoA4NWI8ToF-LYLYIAA) in order to delete the specific plugin I need.
 
## Describe the solution you'd like

I would like a simple way to opt out of the babel plugins. I think the best way is to simply attach the name of the plugin to each plugin, so that the user can filter in a babel.config.js:

```js
// note: this is just my proposal - this doesnt actually work bc plugins arent self identifying
const nextbabel = require("next/babel");

const wrapNextBabel = (api) => {
  const temp = nextbabel(api);
  temp.plugins = templ.plugins.filter(plugin => plugin.name !== 'styled-jsx')
  return temp;
};
return {
  presets: [wrapNextBabel],
  plugins: [],
};

```

(btw - Next.js' typescript fails when you use `babel.config.js` as oppose to `.babelrc` - i ended up routing `.babelrc` to a `babel-preset.js` in order to trick Next.js)

## Describe alternatives you've considered

maintaining this ugly ass fork haha

**Comment by swyxio** [comment]
i think this is a pretty easy PR. if you approve the idea i would be happy to PR> basically just attaching string names to each babel plugin.

**Comment by timneutkens** [comment]
I think the better solution would be for the styled-jsx plugin to correctly no-op when you don't use styled-jsx 🤔

**Comment by swyxio** [comment]
ok.. that one might be a bit above my ability haha. 

i was just trying to generalize from my bad experience - if i have to opt out of this plugin, i might need to opt out of any of the other plugins - so i'd like `next/babel` to help me.

but i think its *also* good to no-op styled-jsx when it is not used. could potentially save a lot of kb in wasteful generated classnames.

**Comment by timneutkens** [comment]
> but i think its also good to no-op styled-jsx when it is not used. could potentially save a lot of kb in wasteful generated classnames.

This should be the case already if you don't use styled-jsx though. Maybe you have a `<style jsx>` somewhere 🤔

**Comment by swyxio** [comment]
ahh you're right. i accidentally left some in there when i copied some boilerplate.
